File: ReadingTool.Site/Controllers/AjaxController.cs
﻿#region License
// AjaxController.cs is part of ReadingTool.Site
// 
// ReadingTool.Site is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// ReadingTool.Site is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with ReadingTool.Site. If not, see <http://www.gnu.org/licenses/>.
// 
// Copyright (C) 2013 Travis Watt
#endregion

using System;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.Mvc;
using ReadingTool.Common;
using ReadingTool.Common.Extensions;
using ReadingTool.Entities;
using ReadingTool.Repository;
using ReadingTool.Services;
using ReadingTool.Site.Attributes;
using ReadingTool.Site.Models.Ajax;

namespace ReadingTool.Site.Controllers.Home
{
    [Authorize]
    [NeedsPersistence]
    public class AjaxController : BaseController
    {
        private readonly Repository<Language> _languageRepository;
        private readonly Repository<Text> _textRepository;
        private readonly Repository<User> _userRepository;
        private readonly Repository<Term> _termRepository;
        private readonly Repository<Tag> _tagRepository;
        private readonly ITextService _textService;

        public AjaxController(
            Repository<Language> languageRepository,
            Repository<Text> textRepository,
            Repository<User> userRepository,
            Repository<Term> termRepository,
            Repository<Tag> tagRepository,
            ITextService textService
            )
        {
            _languageRepository = languageRepository;
            _textRepository = textRepository;
            _userRepository = userRepository;
            _termRepository = termRepository;
            _tagRepository = tagRepository;
            _textService = textService;
        }

        [AjaxRoute]
        [HttpPost]
        public JsonNetResult ResetTerm(Guid termId)
        {
            var term = _termRepository.FindOne(x => x.TermId == termId && x.User == _userRepository.LoadOne(UserId));

            if(term == null)
            {
                throw new Exception("Term not found");
            }

            term.Box = 1;
            term.State = TermState.NotKnown;
            _termRepository.Save(term);

            return new JsonNetResult()
            {
                Data = new TermModel
                {
                    BasePhrase = term.BasePhrase,
                    Definition = term.Definition,
                    Phrase = term.Phrase.ToLowerInvariant(),
                    Sentence = term.Sentence,
                    State = term.State.ToString(),
                    Tags = Tags.ToString(term.Tags.Select(x => x.TagTerm)),
                    TermId = term.TermId,
                    Message = "Term updated to <strong>" + term.State.ToDescription() + " in box 1</strong>",
                    Box = term.Box,
                    StateClass = Term.TermStateToClass(term.State)
                }
            };
        }

        [AjaxRoute]
        [HttpPost]
        public JsonNetResult SaveTerm(SaveTermModel model)
        {
            if(!ModelState.IsValid)
            {
                return new JsonNetResult()
                {
                    Data = new TermModel
                    {
                        State = "failed",
                        Message = "<strong>WARNING</strong>: Term did not save"
                    }
                };
            }

            Term term = null;
            string message;
            short length = (short)model.Phrase.Trim().Split(' ').Length;

            if(model.TermId == null || model.TermId == Guid.Empty)
            {
                term = new Term()
                    {
                        Created = DateTime.Now,
                        Modified = DateTime.Now,
                        BasePhrase = model.BasePhrase.Trim(),
                        Box = 1,
                        Definition = model.Definition.Trim(),
                        Language = _languageRepository.LoadOne(model.LanguageId),
                        User = _userRepository.LoadOne(UserId),
                        Phrase = model.Phrase.Trim(),
                        Sentence = model.Sentence.Trim(),
                        State = (TermState)Enum.Parse(typeof(TermState), model.State, true),
                        Text = _textRepository.LoadOne(model.TextId),
                        Length = length
                    };

                var stateUpdate = Term.NextReviewDate(term);
                term.State = stateUpdate.Item1;
                term.NextReview = stateUpdate.Item2;

                message = "Term created, state is <strong>" + term.State.ToDescription() + "</strong>";
            }
            else
            {
                term = _termRepository.FindOne(x => x.TermId == model.TermId && x.User == _userRepository.LoadOne(UserId));

                if(term == null)
                {
                    throw new Exception("Invalid term");
                }

                term.BasePhrase = model.BasePhrase.Trim();
                term.Sentence = model.Sentence.Trim();
                term.Definition = model.Definition.Trim();
                term.Length = length;

                var newState = (TermState)Enum.Parse(typeof(TermState), model.State, true);
                bool resetBox = false;
                if(term.State != newState && newState == TermState.NotKnown)
                {
                    resetBox = true;
                    term.Box = 1;
                }

                term.State = newState;
                message = "Term updated, state is <strong>" + term.State.ToDescription() + "</strong>";

                if(resetBox)
                {
                    message += ", reset to <strong>box 1</strong>";
                }

                term.Tags.Clear();
                foreach(var tag in Tags.ToTags(model.Tags))
                {
                    var existing = _tagRepository.FindOne(x => x.TagTerm.Equals(tag.ToLowerInvariant()));

                    if(existing == null)
                    {
                        existing = new Tag()
                        {
                            TagTerm = tag
                        };
                    }

                    term.Tags.Add(existing);
                }

                //term.HasTags = term.Tags.Count > 0;

                var stateUpdate = Term.NextReviewDate(term);
                term.State = stateUpdate.Item1;
                term.NextReview = stateUpdate.Item2;
            }

            _termRepository.Save(term);

            return new JsonNetResult()
            {
                Data = new TermModel
                {
                    BasePhrase = term.BasePhrase,
                    Definition = term.Definition,
                    Phrase = term.Phrase.ToLowerInvariant(),
                    Sentence = term.Sentence,
                    State = term.State.ToString(),
                    Tags = Tags.ToString(term.Tags.Select(x => x.TagTerm)),
                    TermId = term.TermId,
                    Message = message,
                    Box = term.Box,
                    StateClass = Term.TermStateToClass(term.State),
                    Length = term.Length
                }
            };
        }

        [AjaxRoute]
        [HttpPost]
        public JsonNetResult FindTerm(Guid? termId, string spanTerm, Guid languageId)
        {
            spanTerm = (spanTerm ?? "").Trim();
            Term term = null;

            short length = (short)spanTerm.Split(' ').Length;

            if(termId == null || termId == Guid.Empty)
            {

                term = _termRepository.FindAll(x =>
                                                   x.PhraseLower.Equals(spanTerm.ToLowerInvariant()) &&
                                                   x.User == _userRepository.LoadOne(UserId) &&
                                                   x.Language == _languageRepository.LoadOne(languageId)
                                                   )
                    .FirstOrDefault();
            }
            else
            {
                term = _termRepository.FindOne(x => x.TermId == termId && x.User == _userRepository.LoadOne(UserId));
            }

            if(term == null)
            {
                return new JsonNetResult()
                    {
                        Data = new TermModel
                            {
                                Phrase = spanTerm.Trim(),
                                Length = length,
                                State = TermState.NotKnown.ToString(),
                                Message = "New word, default to <strong>UNKNOWN</strong>"
                            }
                    };
            }

            return new JsonNetResult()
                {
                    Data = new TermModel
                        {
                            BasePhrase = term.BasePhrase,
                            Definition = term.Definition,
                            Phrase = term.Phrase,
                            Sentence = term.Sentence,
                            State = term.State.ToString(),
                            Tags = Tags.ToString(term.Tags.Select(x => x.TagTerm)),
                            Box = term.Box,
                            TermId = term.TermId,
                            Message = "Current box : <strong>" + term.Box + "</strong>",
                            Length = term.Length
                        }
                };
        }

        [HttpPost]
        [AjaxRoute]
        public JsonNetResult MarkRemaingAsKnown(Guid languageId, Guid textId, string[] terms)
        {
            if(terms == null || terms.Length == 0)
            {
                return new JsonNetResult() { Data = "OK" };
            }

            try
            {
                var language = _languageRepository.FindOne(languageId);
                var user = _userRepository.LoadOne(UserId);
                var text = _textRepository.LoadOne(textId);

                Regex regex = new Regex(@"([" + language.Settings.RegexWordCharacters + @"])", RegexOptions.Compiled);

                foreach(var term in terms.Distinct(StringComparer.Create(CultureInfo.InvariantCulture, true)))
                {
                    if(string.IsNullOrEmpty(term))
                    {
                        continue;
                    }

                    if(!regex.IsMatch(term))
                    {
                        continue;
                    }

                    _termRepository.Save(new Term()
                    {
                        Created = DateTime.Now,
                        Modified = DateTime.Now,
                        Box = 1,
                        Language = language,
                        User = user,
                        Phrase = term,
                        State = TermState.Known,
                        Text = text,
                    });
                }
            }
            catch(Exception e)
            {
                return new JsonNetResult() { Data = e.Message };
            }

            return new JsonNetResult() { Data = "OK" };
        }

        [AjaxRoute]
        [HttpPost]
        public JsonResult EncodeTerm(Guid languageId, Guid dictionaryId, string input)
        {
            try
            {
                var language = _languageRepository.FindOne(languageId);

                if(language == null)
                {
                    throw new Exception("Language is null");
                }

                var dictionary = language.Dictionaries.FirstOrDefault(x => x.DictionaryId == dictionaryId);

                if(dictionary == null)
                {
                    throw new Exception("Dictionary is null");
                }

                var encoder = Encoding.GetEncoding(dictionary.Encoding);
                string encoded = HttpUtility.UrlEncode(input, encoder);
                string result = dictionary.Url.Replace("###", encoded);

                return new JsonNetResult() { Data = result };
            }
            catch(Exception e)
            {
                Elmah.ErrorSignal.FromCurrentContext().Raise(e);
                return new JsonNetResult() { Data = "" };
            }
        }
    }
}

File: app/controllers/AjaxController.php
<?php

use RT\Core\FlashMessage;
use App\Models\Term;
use App\Models\Tag;
use App\Models\Text;
use App\Models\Language;
use RT\Services\ITextService;
use RT\Services\ITermService;
use RT\Services\ILanguageService;
use RT\Services\IParserService;

class AjaxController extends BaseController {
    private $textService;
    private $termService;
    private $parserService;
    private $languageService;
    
    public function __construct(ITextService $textService, ITermService $termService, IParserService $parserService, ILanguageService $languageService) {
        $this->textService = $textService;
        $this->termService = $termService;
        $this->parserService = $parserService;
        $this->languageService = $languageService;
    }
    
    private function termToStateClass($state) {
        $state = trim(strtolower($state));
        
        switch($state) {
            case "known": return "_k";
            case "ignore": return "_i";
            case "ignored": return "_i";
            case "notknown": return "_u";
            case "unknown": return "_u";
            case "notseen": return "_n";
            default: throw new \Exception("Unknown state: " . $state);
        }
    }
    
    private function termStateToDesription($state) {
        $state = trim(strtolower($state));
        
        switch($state) {
            case "known": return "Known";
            case "ignore": return "Ignored";
            case "ignored": return "Ignored";
            case "notknown": return "Unknown";
            case "unknown": return "Unknown";
            case "notseen": return "Not Seen";
            default: throw new \Exception("Unknown state: " . $state);
        }
    }
    
    public function postResetTerm() {
        $termId = Input::get('termId');
        
        $term = $this->termService->find($termId);
        
        if($term==null) {
            return Response::json();
        }
        
        $term->state = "unknown";
        $term->save();
        
        return Response::json(
                array(
                    "basePhrase" => $term->basePhrase,
                    "definition" => $term->definition,
                    "phrase" => $term->phraseLower,
                    "sentence" => $term->sentence,
                    "state" => $term->state,
                    "tags" => array(),
                    "termId" => $term->id,
                    "box" => 1,
                    "message" => "Term updated to <strong>" . $this->termStateToDesription($term->state),
                    "stateClass" => $this->termToStateClass($term->state)
                ));
    }
    
    private function termLength($phrase) {
        if(empty($phrase)) {
            return 0;
        }
        
        $ex = explode(' ', trim($phrase));
        
        return sizeof($ex);
    }
    
    public function postSaveTerm() {
        $rules = array(
            'phrase' => 'max:50',
            'basePhrase' => 'max:50',
            'sentence' => 'max:500',
            'definition' => 'max:500'
        );
        
        $validator = Validator::make(Input::all(), $rules);
        
        if($validator->fails()) {
            return Response::json(array('state'=>'failed', 'message'=>'<strong>WARNING</strong>: Term did not save, invalid input'));
        }
        
        $length = $this->termLength(Input::get('phrase'));
        $message = '';
        $termId = Input::get('termId');
        $tags = explode(',', Input::get('tags'));
        $returnTags = array();
        
        if(!isset($termId) || empty($termId)|| $termId<=0) {
            $term = new Term;
            $term->basePhrase = trim(Input::get('basePhrase'));
            $term->definition =trim(Input::get('definition'));
            $term->language_id = Input::get('languageId');
            $term->user_id = Auth::user()->id;
            $term->phrase = trim(Input::get('phrase'));
            $term->phraseLower = mb_strtolower($term->phrase);
            $term->sentence = trim(Input::get('sentence'));
            
            $state = Input::get('state');
            
            switch(strtolower($state)) {
                case "known": $term->state = 'known'; break;
                case "unknown": $term->state = 'unknown'; break;
                case "notseen": $term->state = 'notseen'; break;
                case "ignore": $term->state = 'ignored'; break;
                default: throw new \Exception('Unknown state: ' . $state);
            }
            
            $term->text_id = Input::get('textId');
            $term->length = $length;
            $term->save();
            
            foreach($tags as $tag) {
                $tag = trim(mb_strtolower($tag));
                
                if(empty($tag) || $tag==='') {
                    continue;
                }
                
                if(in_array($tag, $returnTags)) {
                    continue;
                }
                
                $exists = Tag::where('tag','=', $tag)->first();
                
                if($exists==null) {
                    $t = new Tag;
                    $t->tag = $tag;
                    $t->save();
                    
                    $term->Tags()->attach($t->id);
                } else {
                    $term->Tags()->attach($exists->id);
                }
                
                array_push($returnTags, $tag);
            }
            
            //$term->save();

            $message = "Term created, state is <strong>" . $this->termStateToDesription($term->state) . "</strong>";
        } else {
            $term = $this->termService->find($termId);
            
            if($term==null) {
                return Response::json(array('state'=>'failed', 'message'=>'<strong>WARNING</strong>: Term did not save, term not found'));    
            }

            $term->basePhrase = trim(Input::get('basePhrase'));
            $term->definition =trim(Input::get('definition'));
            
            $sentence = trim(Input::get('sentence'));
            
            if(strcmp(trim($term->sentence), $sentence)!=0) {
                $term->sentence = $sentence;
                $term->text_id = Input::get('textId');
            }
            
            $state = Input::get('state');
            
            switch(strtolower($state)) {
                case "known": $term->state = 'known'; break;
                case "unknown": $term->state = 'unknown'; break;
                case "notseen": $term->state = 'notseen'; break;
                case "ignore": $term->state = 'ignored'; break;
                default: throw new \Exception('Unknown state: ' . $state);
            }
            
            $term->Tags()->detach();
            
            foreach($tags as $tag) {
                $tag = trim(mb_strtolower($tag));
                
                if(empty($tag) || $tag==='') {
                    continue;
                }
                
                if(in_array($tag, $returnTags)) {
                    continue;
                }
                
                $exists = Tag::where('tag','=', $tag)->first();
                
                if($exists==null) {
                    $t = new Tag;
                    $t->tag = $tag;
                    $t->save();
                    
                    $term->Tags()->attach($t->id);
                } else {
                    $term->Tags()->attach($exists->id);
                }
                
                array_push($returnTags, $tag);
            }
            
            $term->save();
            
            $message = "Term updated, state is <strong>" . $this->termStateToDesription($term->state) . "</strong>";
        }
        
        return Response::json(
                array(
                    "basePhrase" => $term->basePhrase,
                    "definition" => $term->definition,
                    "phrase" => $term->phraseLower,
                    "sentence" => $term->sentence,
                    "state" => $term->state,
                    "tags" => implode(',', $returnTags),
                    "termId" => $term->id,
                    "message" => $message,
                    "box" => 1,
                    "stateClass" => $this->termToStateClass($term->state),
                    "exists" => true
                ));
    }
    
    public function postFindTerm() {
        $termId = Input::get('termId');
        $languageId = Input::get('languageId');
        $spanTerm = Input::get('spanTerm');
        
        $length = $this->termLength($spanTerm);
        
        $term = null;
        
        if(!isset($termId) || empty($termId) || $termId<=0) {
            $term = $this->termService->findByPhraseAndLanguage($spanTerm, $languageId);
        } else {
            $term = $this->termService->find($termId);
        }
        
        if($term==null) {
            return Response::json(array(
                    'phrase' => $spanTerm,
                    'length' => $length,
                    'state' => 'unknown',
                    'message' => 'New word, default to <strong>UNKNOWN</strong>',
                    'exists' => false
            ));
        } else {
            $tags = array();
            foreach($term->tags as $t) {
                array_push($tags, $t->tag);
            }
            
            return Response::json(array(
                    'basePhrase' => $term->basePhrase,
                    'definition' => $term->definition,
                    'phrase' => $term->phrase,
                    'sentence' => $term->sentence,
                    'state' => $term->state,
                    'termId' => $term->id,
                    'message' => '',
                    'length' => $term->length,
                    'box' => 1,
                    'tags' => implode(',',$tags),
                    'exists' => true
            ));
        }
    }
    
    public function postMarkRemainingAsKnown() {
        $languageId = Input::get('languageId');
        $textId = Input::get('textId');
        $userId = Auth::user()->id;
        $terms = Input::get('terms');
        
        $language = $this->languageService->find($languageId);
        $termTestRegex = "/([^" . $language->getSettings()->RegexWordCharacters . "]+)/";
        
        $processed = array();
        
        foreach($terms as $term) {
            $term = trim($term);
            $lower = mb_strtolower($term);
            
            if(empty($term) || $term==='' || array_key_exists($lower, $processed)) {
                continue;
            }
            
            if(preg_match($termTestRegex, $term)==0) {
                $t = new Term;
                $t->language_id = $languageId;
                $t->user_id = $userId;
                $t->phrase = $term;
                $t->phraseLower = $lower;
                $t->text_id = $textId;
                $t->state = 'known';
                $t->length = 1;
                $t->save();
                
                $processed[$lower] = true;
            }
        }
        
        return "OK";
    }
    
    public function postEncodeTerm() {
        $languageId = Input::get('languageId');
        $dictionaryId = Input::get('dictionaryId');
        $input = Input::get('input');
        
        //$language = $this->languageService->find($languageId);
        $dictionary = $this->languageService->findDictionary($dictionaryId);
        $word = mb_convert_encoding($input, "$dictionary->encoding", 'UTF-8');
        
        return Response::json(str_replace('###', urlencode($word), $dictionary->url));
    }
}
